pkg/kubelet/labels.go

```diff
@@ -0,0 +1,239 @@
+	}
+	err := ls.writeLabelsFile(ls.checkpointFile, clearLabels)
+	if err != nil {
+		return fmt.Errorf("Could not append checkpoint file %v", err)
+	}
+
+	ls.checkpointRequiresAppend = false
+	return nil
+}
+
+// WriteCheckpoint writes the new node labels to the checkpoint file, and clears the old checkpoin labels.
+// Will not write the checkpoint file if has already been written.
+func (ls *labelSet) WriteLabelCheckpoint() error {
+	if !ls.checkpointRequiresWrite {
+		return nil
+	}
+
+	// Write new labels to the checkpoint
+	clearLabels := map[string]string{}
+	for k := range ls.nodeLabelsToAdd {
+		clearLabels[k] = ""
+	}
+	err := ls.writeLabelsFile(ls.checkpointFile, clearLabels)
+	if err != nil {
+		return fmt.Errorf("Could not write checkpoint file %v", err)
+	}
+
+	ls.checkpointRequiresWrite = false
+	return nil
+}
+
+// WriteLabels adds / updates labels read from the label directory to the map.  Removes labels from the map
+// that were in the label directory, but have been removed.
+func (ls *labelSet) WriteLabelsToNodeMeta(node *api.Node) {
+	labels := node.ObjectMeta.Labels
+	for k := range ls.nodeLabelsToClear {
+		delete(labels, k)
+	}
+	for k, v := range ls.nodeLabelsToAdd {
+		labels[k] = v
+	}
+}
+
+// readLabels reads labels from the label directory and returns a map of the results.
+func (ls *labelSet) readLabels() (map[string]string, error) {
+	dstLabels := map[string]string{}
+	if ls.labelDirectory == "" {
+		return dstLabels, nil
+	}
+	files, err := ioutil.ReadDir(ls.labelDirectory)
+	if err != nil {
+		return nil, err
+	}
+
+	// Load each file
+	for _, fi := range files {
+		f := path.Join(ls.labelDirectory, fi.Name())
+		if ext := filepath.Ext(fi.Name()); ext != ".json" && ext != ".yaml" {
+			glog.Warningf("Unsupported file type (must be json or yaml) in node labels directory: %s", f)
+			continue
+		}
+		srcLabels, err := ls.readLabelsFile(f)
+		if err != nil {
+			return nil, err
+		}
+		for k, v := range srcLabels {
+			if _, dk := dstLabels[k]; dk {
+				return nil, fmt.Errorf(
+					"Multiple values found for label key %s.  Check %s for files defining the label.", ls.labelDirectory, k)
+			}
+			dstLabels[k] = v
+		}
+	}
+	return dstLabels, nil
+}
+
+// readCheckpoint reads labels from the checkpoint file and returns a map of the results.
+func (ls *labelSet) readCheckpoint() (map[string]string, error) {
+	if ls.checkpointFile == "" {
+		return map[string]string{}, nil
+	}
+	return ls.readLabelsFile(ls.checkpointFile)
+}
+
+// writeLabelsFile writes the map of labels to a file as json
+func (ls *labelSet) writeLabelsFile(path string, labels map[string]string) error {
+	b, err := json.Marshal(labels)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(path, b, 0644)
+}
+
+// readLabelsFile reads in and parses the yaml or json node labels file
+func (ls *labelSet) readLabelsFile(path string) (map[string]string, error) {
+	labels := make(map[string]string, 0)
+	kps := make(map[string]interface{}, 0)
```

Comment:
This format needs to be formalized. How about using `KubeletConfiguration` Object here?